### PR TITLE
ORC-1152:[C++] Support writing short decimals in RLEv2

### PR DESCRIPTION
**Pull request by coderex2522**
### What changes were proposed in this pull request?
This PR is to support writing short decimal64 in ORCv2. 
The original writer codes come from [ORC-49](https://github.com/apache/orc/pull/257)

### Why are the changes needed?
Before this PR, the cpp writer is unable to write short decimal columns in ORCv2.

### How was this patch tested?
Add the new file version UNSTABLE_PRE_2_0 test in TestWriter.cc.
WriterTest.writeDecimal64Column will test the Decimal64ColumnWriterV2 class.

### CODE REVIEW

**Review by dongjoon-hyun — COMMENTED** [review]
Thank you for your efforts on C++ and Java feature parity, @coderex2522 .

File: c++/test/TestWriter.cc [hunk 1]
```diff
@@ -1996,5 +1996,5 @@ namespace orc {
     }
   }
 
-  INSTANTIATE_TEST_CASE_P(OrcTest, WriterTest, Values(FileVersion::v_0_11(), FileVersion::v_0_12()));
+  INSTANTIATE_TEST_CASE_P(OrcTest, WriterTest, Values(FileVersion::v_0_11(), FileVersion::v_0_12(), FileVersion::UNSTABLE_PRE_2_0()));
```
Comment (wgtmac) [inline c++/test/TestWriter.cc:1999]: Should we generate some ORC files and add some test cases for cross-language compatibility? We have had issues when Java ORC reader cannot understand bloom filter produced by the C++ writer (which is caught by @stiga-huang). We may be safely assured when test cases report that the Java ORC reader can correctly read UNSTABLE_PRE_2_0 file created by the C++ ORC writer and vice versa. @dongjoon-hyun @stiga-huang @coderex2522

Comment (dongjoon-hyun) [reply c++/test/TestWriter.cc:1999]: +1 for having that. BTW, what is the ORC issue ID for that bloom filter compatibility issue?

Comment (wgtmac) [reply c++/test/TestWriter.cc:1999]: > at is the ORC issue ID for that

https://issues.apache.org/jira/browse/ORC-1024

Comment (dongjoon-hyun) [reply c++/test/TestWriter.cc:1999]: Thank you.

Comment (coderex2522) [reply c++/test/TestWriter.cc:1999]: > Should we generate some ORC files and add some test cases for cross-language compatibility? We have had issues when Java ORC reader cannot understand bloom filter produced by the C++ writer (which is caught by @stiga-huang). We may be safely assured when test cases report that the Java ORC reader can correctly read UNSTABLE_PRE_2_0 file created by the C++ ORC writer and vice versa. @dongjoon-hyun @stiga-huang @coderex2522

The [PR](https://github.com/apache/orc/pull/1062) has been tested using c++ ORC reader to read UNSTABLE_PRE_2_0 file created by the java ORC writer, so here only adds the test about java reader reads the file written by c++ writer.

File: java/core/src/java/org/apache/orc/impl/TreeReaderFactory.java [hunk 2]
```diff
@@ -1783,8 +1783,9 @@ protected Decimal64TreeReader(int columnId,
 
     @Override
     public void checkEncoding(OrcProto.ColumnEncoding encoding) throws IOException {
-      if ((encoding.getKind() != OrcProto.ColumnEncoding.Kind.DIRECT)) {
-        throw new IOException("Unknown encoding " + encoding + " in column " +
+      if (encoding.getKind() != OrcProto.ColumnEncoding.Kind.DIRECT &&
+          encoding.getKind() != OrcProto.ColumnEncoding.Kind.DIRECT_V2) {
+        throw new IOException("rexUnknown encoding " + encoding + " in column " +
```
Comment (wgtmac) [inline java/core/src/java/org/apache/orc/impl/TreeReaderFactory.java:1788]: Fix the error message.

Comment (coderex2522) [reply java/core/src/java/org/apache/orc/impl/TreeReaderFactory.java]: Done.

**Review by wgtmac — APPROVED** [review]
+1 LGTM. Thanks @coderex2522 and @dongjoon-hyun!

**Comment by dongjoon-hyun** [comment]
Thank you, @wgtmac . Could you resolve ORC-1152 correspondingly?

**Pull request merged**